### PR TITLE
Track the user's relay list using polling

### PR DESCRIPTION
**Pull request by ksedgwic**
This is the third attempt, the previous were #346 and #380

This version uses polling instead of async operations

### CODE REVIEW

**Comment by jb55** [comment]
seeing concerning warnings on this branch:

```
2024-11-22T20:37:47.056689Z  WARN notedeck::app: got unknown eose[..] ... 97 tracked subscriptions
```

also:

```
2024-11-22T20:37:50.812563Z ERROR notedeck::app: wss://nostr.africa/: IO error: failed to lookup address information: nodename nor servname provided, or not known
2024-11-22T20:37:50.812570Z ERROR notedeck::app: wss://relay.current.fyi/: IO error: failed to lookup address information: nodename nor servname provided, or not known
2024-11-22T20:37:50.812575Z ERROR notedeck::app: wss://relayer.fiatjaf.com/: IO error: failed to lookup address information: nodename nor servname provided, or not known
2024-11-22T20:37:50.812581Z ERROR notedeck::app: wss://rsslay.fiatjaf.com/: IO error: failed to lookup address information: nodename nor servname provided, or not known
2024-11-22T20:37:50.958067Z ERROR notedeck::app: wss://nostr.zebedee.cloud/: IO error: invalid peer certificate contents: invalid peer certificate: CertNotValidForName
2024-11-22T20:37:51.124584Z ERROR notedeck::app: wss://relay.plebstr.com/: IO error: invalid peer certificate contents: invalid peer certificate: CertExpired
2024-11-22T20:37:51.141660Z ERROR notedeck::app: wss://nostr.onsats.org/: IO error: invalid peer certificate contents: invalid peer certificate: CertNotValidForName
```

which seems to be trying to fetch from a bunch of relays I don't recognize, is this trying to do the outbox model?

**Comment by jb55** [comment]
there also seems to be log entries that are creating newlines which we should try to avoid (Muted, desired_relays)

**Comment by ksedgwic** [comment]
> outbox model?

Nope, just k:10002

Can you try with `RUST_LOG=error,notedeck::accounts=debug,notedeck::muted=debug`

**Comment by ksedgwic** [comment]
Which npub are you using?  I'll see if I can reproduce ...

**Comment by ksedgwic** [comment]
> creating newlines
fixed

File: src/accounts/mod.rs [hunk 1]
```diff
@@ -17,18 +21,118 @@ use crate::{
...
+                        Some("alt") => {
+                            // ignore for now
+                        }
+                        Some(x) => {
+                            error!("harvest_nip65_relays: unexpected tag type: {}", x);
+                        }
+                        None => {
+                            error!(
+                                "harvest_nip65_relays: malformed tag value: {:?}",
+                                tag.get_unchecked(0).variant()
```
Comment (jb55) [inline src/accounts/mod.rs:113]: this would crash on an empty tag

File: src/muted.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            return true;
+        }
+        // FIXME - Implement hashtag muting here
+        let content = note.content().to_lowercase();
+        for word in &self.words {
+            if content.contains(&word.to_lowercase()) {
+                debug!("{}: MUTED word: {}", hex::encode(note.id()), word);
+                return true;
+            }
+        }
```
Comment (jb55) [inline src/muted.rs:51]: let's not add this for now, we will likely need to have an optimized data structure in nostrdb to properly mute words. this mutes substrings which is not ideal.

File: src/muted.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+impl Muted {
+    pub fn is_muted(&self, note: &Note) -> bool {
+        if self.pubkeys.contains(note.pubkey()) {
+            debug!(
+                "{}: MUTED pubkey: {}",
+                hex::encode(note.id()),
+                hex::encode(note.pubkey())
+            );
+            return true;
```
Comment (jb55) [inline src/muted.rs:42]: we probably don't want to hex encode for every is_muted check. these will be evaluated regardless of the log being enabled.

File: src/muted.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        // FIXME - Implement hashtag muting here
+        let content = note.content().to_lowercase();
+        for word in &self.words {
+            if content.contains(&word.to_lowercase()) {
+                debug!("{}: MUTED word: {}", hex::encode(note.id()), word);
```
Comment (jb55) [inline src/muted.rs:48]: this too

File: src/accounts/mod.rs [hunk 5]
```diff
@@ -119,8 +120,98 @@ impl AccountRelayData {
...
+            .map(|qr| qr.note_key)
+            .collect::<Vec<NoteKey>>();
+        let muted = Self::harvest_nip51_muted(ndb, &txn, &nks);
+        debug!("pubkey {}: initial muted {:?}", hex::encode(pubkey), muted);
+
+        // Id for future remote relay subscriptions
+        let subid = Uuid::new_v4().to_string();
+
+        // Add remote subscription to existing relays
+        pool.subscribe(subid.clone(), vec![filter.clone()]);
```
Comment (jb55) [inline src/accounts/mod.rs:161]: we should coordinate this subscription elsewhere, but this is fine for now. I will add a TODO

**Review by jb55 — CHANGES_REQUESTED** [review]
just a few things then we're good to go!

File: src/muted.rs (hunk 4 above)
Comment (ksedgwic) [reply src/muted.rs:48]: also only encoded when it matches (and it's commented out for now)

File: src/account_manager.rs [hunk 6]
```diff
@@ -97,18 +117,40 @@ pub fn process_accounts_view_response(
...
         };
 
         let currently_selected_account = get_selected_index(&accounts, &key_store);
+        let account_data = BTreeMap::new();
+        let forced_relays: BTreeSet<String> = forced_relays
+            .into_iter()
+            .map(|u| Self::canonicalize_url(&u))
+            .collect();
+        let bootstrap_relays = [
+            "ws://localhost:8080",
```
[the remark this answers is not in the document]
Comment (ksedgwic) [reply src/account_manager.rs]: removed

File: src/muted.rs (hunk 2 above)
Comment (ksedgwic) [reply src/muted.rs]: commented out w/ reason

File: src/accounts/mod.rs (hunk 1 above)
Comment (ksedgwic) [reply src/accounts/mod.rs]: fixed

**Comment by jb55** [comment]
sorry for the delay! thanks!

**Pull request merged**